### PR TITLE
Improve LSP responsiveness by always destroying ASTs in parallel during cancelation

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Improve LSP responsiveness by always destroying ASTs in parallel during cancelation.

When LSP is typechecking the world and a new edit comes in that causes LSP to cancel that typecheck, the code used to destroy all ASTs on a single thread, which is very slow on large projects and resulted in a noticeable pause.

This change enforces that cancelation always destroys trees in parallel by making the only way to construct a `ast::ParsedFilesOrCancelled` object in the "cancelled" state is to call a new static function that destroys the trees passed to it in parallel.

Note that `pipeline::typecheck` always destroys ASTs in parallel after #4952 landed, so this change only impacts `namer` and `resolver`.

One negative of the way that I implemented this is that `ast/Trees.cc` now contains WorkerPool-related code to implement `ast::ParsedFilesOrCancelled::cancel`. I'd be happy to entertain alternative code organization ideas.

### Motivation

Improve Sorbet LSP responsiveness.

### Test plan

Cancelation is tested by existing tests. I also manually tested it.

### CODE REVIEW

File: ast/Trees.cc [hunk 1]
```diff
@@ -1411,6 +1414,31 @@ string BlockArg::nodeName() {
 ParsedFilesOrCancelled::ParsedFilesOrCancelled() : trees(nullopt){};
 ParsedFilesOrCancelled::ParsedFilesOrCancelled(std::vector<ParsedFile> &&trees) : trees(move(trees)) {}
 
+ParsedFilesOrCancelled ParsedFilesOrCancelled::cancel(std::vector<ParsedFile> &&trees, WorkerPool &workers) {
+    if (!trees.empty()) {
+        absl::BlockingCounter threadBarrier(std::max(workers.size(), 1));
```
Comment (jvilk-stripe) [inline ast/Trees.cc:1419]: Note: I'm pretty sure `workers.size()` returns 0 when threading is disabled, hence the `std::max` with `1`. I can add a comment if desired.

File: ast/Trees.h [hunk 2]
```diff
@@ -235,11 +239,16 @@ struct ParsedFile {
 class ParsedFilesOrCancelled final {
 private:
     std::optional<std::vector<ParsedFile>> trees;
+    ParsedFilesOrCancelled();
```
Comment (jvilk-stripe) [inline ast/Trees.h:242]: I made this private to enforce that callsites must destroy trees in parallel, in the hopes of preventing a regression in the future.

File: ast/Trees.cc (hunk 1 above)
Comment (aprocter-stripe) [reply ast/Trees.cc:1419]: Comment wouldn't hurt, I spose!

File: ast/Trees.cc [hunk 3]
```diff
@@ -1411,6 +1414,31 @@ string BlockArg::nodeName() {
...
+        auto fileq = make_shared<ConcurrentBoundedQueue<ast::ParsedFile>>(trees.size());
+        for (auto &tree : trees) {
+            fileq->push(move(tree), 1);
+        }
+
+        workers.multiplexJob("deleteTrees", [fileq, &threadBarrier]() {
+            {
+                ast::ParsedFile job;
+                for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
+                    // Do nothing; allow the destructor of `ast::ParsedFile` to run.
```
Comment (aprocter-stripe) [inline ast/Trees.cc:1429]:
```suggestion
                    // Do nothing; allow the destructor of `ast::ParsedFile` to run for `job`.
```

**Review by aprocter-stripe — APPROVED** [review]
LGTM! Just a couple of non-blocking cosmetic suggestions.

I don't find the introduction of `WorkerPool`-related code into `Trees.cc` all that bothersome, but not sure if anyone else will have stronger opinions.

**Pull request merged**